request_handlers.go

```diff
@@ -0,0 +1,212 @@
+package main
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"strings"
+	"text/template"
+	"time"
+)
+
+func ShowRoot(w http.ResponseWriter, r *http.Request) {
+	logHandlerIntro(r.Method, r.URL.Path, r.Form)
+	renderPage(w, r, "templates/index.tmpl", "Home")
+}
+
+func ShowRegistration(w http.ResponseWriter, r *http.Request) {
+	logHandlerIntro(r.Method, r.URL.Path, r.Form)
+	renderPage(w, r, "templates/registration.tmpl", "Registration")
+}
+
+func ShowLogin(w http.ResponseWriter, r *http.Request) {
+	logHandlerIntro(r.Method, r.URL.Path, r.Form)
+	renderPage(w, r, "templates/login.tmpl", "Login")
+}
+
+func CreateUser(w http.ResponseWriter, r *http.Request) {
+	r.ParseForm()
+	logHandlerIntro(r.Method, r.URL.Path, r.Form)
+
+	newUser := User{
+		Username: r.Form["username"][0],
+		Password: r.Form["password"][0],
+	}
+
+	row := db.QueryRowx(
+		`INSERT INTO users (username, password) VALUES ($1, $2) RETURNING id`,
+		newUser.Username,
+		newUser.Password,
+	)
+	err := row.Scan(&newUser.Id)
+	if err != nil {
+		log.Println(err)
+		http.Error(w, "Error registering", http.StatusInternalServerError)
+		return
+	}
+
+	log.Printf("User added: %+v\n", newUser)
+	fmt.Fprintln(w, "Success!")
+}
+
+func Login(w http.ResponseWriter, r *http.Request) {
+	r.ParseForm()
+	logHandlerIntro(r.Method, r.URL.Path, r.Form)
+
+	user := User{
+		Username: r.Form["username"][0],
+		Password: r.Form["password"][0],
+	}
+
+	// Find the user.
+	err := db.Get(
+		&user,
+		`SELECT * FROM users WHERE username = $1 AND password = $2`,
+		user.Username,
+		user.Password,
+	)
+	if err != nil || !user.Id.Valid {
+		log.Println(err)
+		http.Error(w, "Error logging in", http.StatusBadRequest)
+		return
+	}
+
+	// Already checked the case where this fails above.
+	userId, _ := user.Id.Value()
+	authToken, err := findLoginAuthToken(userId.(string))
+
+	if err == nil {
+		// We have an auth token that is still valid, let's use that.
+		_, err = db.Exec(
+			`UPDATE authentications SET updated_at = $1 WHERE token = $2 AND deleted_at IS NULL`,
+			time.Now(),
+			authToken,
+		)
+		if err != nil {
+			log.Println(err)
+			http.Error(w, "Error", http.StatusInternalServerError)
+		}
+	} else {
+		// Create a new authentication for this user.
+		row := db.QueryRowx(`INSERT INTO authentications (user_id) VALUES ($1) RETURNING token`, userId)
+		err = row.Scan(&authToken)
+		if err != nil {
+			log.Println(err)
+			http.Error(w, "Error logging in", http.StatusInternalServerError)
+			return
+		}
+	}
+
+	// Invalidate all outstanding authentications for this user.
```

Comment:
(Noob Question) Is this how flushing authentication tokens normally works? Wouldn't you eventually accumulate a lot of `deleted = true` authentication tokens and make queries slower?

Reply:
I don't know.
I figure leaving authentications in for audit trails might be a worthwhile thing.
I *do* need to add appropriate indices in, though.